--- alsa-utils-1.1.0/alsaloop/Makefile.am.alsaloop-static	2015-10-27 17:34:26.000000000 +0100
+++ alsa-utils-1.1.0/alsaloop/Makefile.am	2016-01-13 14:31:00.799717433 +0100
@@ -1,9 +1,9 @@
 AM_CPPFLAGS = -I$(top_srcdir)/include
 LDADD = -lm
-AM_CFLAGS = -D_GNU_SOURCE
-if HAVE_SAMPLERATE
-LDADD += -lsamplerate
-endif
+AM_CFLAGS = -D_GNU_SOURCE -DHAVE_SAMPLERATE_H
+#if HAVE_SAMPLERATE
+LDADD += -L. -lsamplerate
+#endif
 # LDFLAGS = -static
 # CFLAGS += -g -Wall
 
